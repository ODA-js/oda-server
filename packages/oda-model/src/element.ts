import { IValidate, IValidator, IValidationResultInput } from './validation';
import { merge } from 'lodash';
import { MetaModelType, Nullable, assignValue } from './types';

export const Internal = Symbol.for('internal');
export const MetaData = Symbol.for('metadata');

export interface ElementMetaInfo {}

export interface ElementInternal<T extends ElementMetaInfo> {
  [MetaData]: T;
}

export interface ElementInput<T extends ElementMetaInfo> {
  metadata?: Partial<T>;
}

export interface ElementOutput<T extends ElementMetaInfo> {
  metadata: T;
}

export interface IUpdatableBase {
  /**
   * update item with data
   * @param payload the update payload
   */
  updateWith(payload: any): void;
  /**
   * return copy of object item that is suitable for creating new one
   */
  toObject(): any;
}

/**
 * updatable items
 */
export interface IUpdatable<
  T extends ElementMetaInfo,
  I extends ElementInput<T>,
  O extends ElementOutput<T>
> extends IUpdatableBase {
  /**
   * update item with data
   * @param payload the update payload
   */
  updateWith(payload: Nullable<I>): void;
  /**
   * return copy of object item that is suitable for creating new one
   */
  toObject(): O;
}

/**
 * Meta interface for updating items with metadata
 */
export interface IMeta<
  M extends ElementMetaInfo,
  I extends ElementInput<M>,
  O extends ElementOutput<M>
> extends IUpdatable<M, I, O>, IValidate {
  /**
   * the kind of current item
   */
  readonly modelType: MetaModelType;
  /**
   * meta information that is outside of model notations and can be customized as well
   */
  readonly metadata: Readonly<M>;
}

const defaultMetaInfo = {};
const defaultInput = { metadata: defaultMetaInfo };

export class Element<
  M extends ElementMetaInfo,
  I extends ElementInput<M>,
  S extends ElementInternal<M>,
  O extends ElementOutput<M>
> implements IMeta<M, I, O> {
  public get modelType(): MetaModelType {
    return 'element';
  }
  public [Internal]: S;
  /**
   * metadata is not immutable
   */
  public get metadata(): M {
    return this[Internal][MetaData];
  }

  public validate(validator: IValidator): IValidationResultInput[] {
    return validator.check(this);
  }

  constructor(init: ElementInput<M>) {
    this[Internal] = {} as S;
    this.updateWith(merge({}, defaultInput, init));
  }
  /**
   * patches metadata for entity
   * @param input metadata patch object
   */
  public updateWith(input: Nullable<ElementInput<M>>) {
    assignValue<Element<M, I, S, O>, ElementInput<M>, ElementInternal<M>>({
      src: this,
      input,
      field: 'metadata',
      effect: (_, value) =>
        (this[Internal][MetaData] = merge({}, this[Internal][MetaData], value)),
      setDefault: _ =>
        (this[Internal][MetaData] = merge({}, defaultMetaInfo) as M),
    });
  }
  /**
   * make item usable within another constructor
   * it is suitable for cloning items
   */
  public toObject(): O {
    return merge({}, { metadata: this.metadata } as O);
  }
}
