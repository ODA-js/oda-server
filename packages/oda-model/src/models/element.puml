@startuml

interface IMeta
interface IUpdatable
interface IValidate

IUpdatable <|-- IMeta
IValidate <|-- IMeta

class Element

IMeta <|-- Element


@enduml
