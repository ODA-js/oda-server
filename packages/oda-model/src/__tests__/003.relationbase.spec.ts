import 'jest';
import {
  RelationBase,
  relationBaseDefaultMetaInfo,
  relationBaseDefaultInput,
} from '../relationbase';
import { MetaData } from '../element';

describe('RelationBase', () => {
  it('creates', () => {
    const res = new RelationBase({});
    expect(res[MetaData]).toMatchObject(relationBaseDefaultMetaInfo);
    expect(res.toObject()).toMatchObject(relationBaseDefaultInput);
    expect(res.toObject()).toMatchSnapshot('toObject');
  });
  it('create base fields', () => {
    const res = new RelationBase({
      name: 'A',
      embedded: true,
      entity: 'AA',
      field: 'F',
      opposite: 'F',
    });
    expect(res.name).toBe('A');
    expect(res.embedded).toBe(true);
    expect(res.entity).toBe('AA');
    expect(res.field).toBe('F');
    expect(res.opposite).toBe('f');
  });

  it('update Name', () => {
    debugger;
    const res = new RelationBase({
      embedded: true,
      entity: 'AA',
      field: 'F',
      opposite: 'F',
    });
    expect(res.name).toBe('AARelatedToFs');
    expect(res.fullName).toBe('AARelatedToFs');
    expect(res.shortName).toBe('Fs');
    expect(res.normalName).toBe('AAFs');
    expect(res.embedded).toBe(true);
    expect(res.entity).toBe('AA');
    expect(res.field).toBe('F');
    expect(res.opposite).toBe('f');
    expect(res.verb).toBe('RelatedTo');
    expect(res.single).toBe(undefined);
    expect(res.stored).toBe(undefined);
    expect(() => res.ref).toThrow();
  });
  it('opposite is settable', () => {
    debugger;
    const res = new RelationBase({
      embedded: true,
      entity: 'AA',
      field: 'F',
    });
    expect(res.opposite).toBe(undefined);
    expect(() => (res.opposite = 'F')).not.toThrow();
    expect(res.opposite).toBe('f');
  });
});
