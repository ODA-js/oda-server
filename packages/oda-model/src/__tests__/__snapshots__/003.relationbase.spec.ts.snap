// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelationBase creates: toObject 1`] = `
Object {
  "embedded": false,
  "entity": undefined,
  "field": undefined,
  "metadata": Object {
    "name": Object {
      "full": "RelatedTo",
      "normal": "",
      "short": "",
    },
    "persistence": Object {
      "embedded": false,
    },
    "verb": "RelatedTo",
  },
  "name": "RelatedTo",
  "opposite": undefined,
}
`;
